src/tqec/compile/substitute.py

```diff
@@ -0,0 +1,218 @@
+    assert (
+        not key.spec1.is_spatial_junction and not key.spec2.is_spatial_junction
+    ), "Both cubes must not be spatial junction."
+    assert pipe_type.direction != Direction3D.Z, "Pipe direction must be X or Y."
+    # TODO: If the pipe has a hadamard, apply the spatial hadamard transformation
+    if pipe_type.has_hadamard:
+        raise TQECException("Hadamard on spatial pipes is not supported yet.")
+    substitute_side1 = (
+        PlaquetteSide.RIGHT
+        if pipe_type.direction == Direction3D.X
+        else PlaquetteSide.DOWN
+    )
+    substitute_side2 = substitute_side1.opposite()
+    # Get reset/mesurement basis
+    basis = pipe_type.value[2].upper()
+    block1 = _substitute_on_the_border(block1, substitute_side1, basis)
+    block2 = _substitute_on_the_border(block2, substitute_side2, basis)
+    return block1, block2
+
+
+def _substitute_on_the_border(
+    block: CompiledBlock,
+    plaquette_side: PlaquetteSide,
+    basis: str,
+) -> CompiledBlock:
+    if plaquette_side == PlaquetteSide.LEFT:
+        substitution = {1: 6, 8: 9, 7: 10}
+    elif plaquette_side == PlaquetteSide.UP:
+        substitution = {2: 12, 5: 10, 6: 9}
+    elif plaquette_side == PlaquetteSide.DOWN:
+        substitution = {3: 7, 14: 10, 13: 9}
+    else:  # plaquette_side == PlaquetteSide.RIGHT:
+        substitution = {4: 13, 11: 9, 12: 10}
+    # 1. Fill in/Substitue the plaquettes on the border
+    for i in range(block.num_layers):
+        block = block.with_updated_layer(
+            plaquettes_to_update={
+                dst: deepcopy(block.layers[i][src]) for dst, src in substitution.items()
+            },
+            layer_to_update=i,
+        )
+    # 2. Add resets/measurements on the middle of the border
+    for i in substitution.keys():
+        _inplace_add_resets_on_data_qubits(
+            block.layers[0][i], plaquette_side, ResetBasis(basis)
+        )
+        _inplace_add_measurements_on_data_qubits(
+            block.layers[-1][i], plaquette_side, MeasurementBasis(basis)
+        )
+
+    return block
+
+
+def _inplace_add_resets_on_data_qubits(
+    plaquette: Plaquette,
+    side: PlaquetteSide,
+    reset_basis: ResetBasis = ResetBasis.Z,
+    reset_moment_schedule: int = 1,
+) -> None:
+    qubits = plaquette.qubits.get_qubits_on_side(side)
+
+    # Do not add resets on qubits that have already be reset
+    reset_moment = plaquette.circuit.moment_at_schedule(reset_moment_schedule)
+    reset_to_add = []
+    for q in qubits:
+        gq = q.to_grid_qubit()
+        if reset_moment.operates_on([gq]):
+            continue
+        reset_to_add.append(reset_basis(gq))
+
+    plaquette.circuit.add_to_schedule_index(
+        reset_moment_schedule, cirq.Moment(*reset_to_add)
+    )
+
+
+def _inplace_add_measurements_on_data_qubits(
+    plaquette: Plaquette,
+    side: PlaquetteSide,
+    measurement_basis: MeasurementBasis = MeasurementBasis.Z,
+    measurement_moment_schedule: int = 8,
```

Comment:
This hard-coded value is a call for subtle bugs as soon as we will use different plaquettes (e.g., plaquettes that require to execute CNOTs on 5 timesteps instead of 4 to keep the logical qubits compatible in the case of a spatial L-shaped junction).

Idealy, `ScheduledCircuit.add_to_schedule_index` should support negative indexing (e.g., -1 means "last schedule" and -2 means "last schedule - 1") and this default value should be -1.